### ISSUE TITLE
Define Out-of-Band Information

### ISSUE THREAD

**Issue by mmccool** [w3c/wot-discovery#30]
Assuming directories cannot modify the TDs they store, what out-of-band information should a directory associate with each TD in its information model?   My thought is that this information should be subject to query as well, so TDs are just part of each "record" in the directory's database.  Here are some candidates:
- local id.   In case the TD does not contain a global id, the database should be assigning a local id to be used to retrieve a specific TD.
- timestamps.  Date of first creation, date of last update
- ttl.  how long before the TD is automatically deleted.

**Comment by benfrancis** [comment]
I would suggest any metadata like this should be stored in the TD itself, or via the protocol the TD is served by (e.g. Expired and Last-Modified headers of HTTP). 

I wouldn't think the internal database IDs of Thing Descriptions should be exposed to WoT clients by directories, the Thing Description URI (or ID member) should be sufficient.

All a directory really needs to serve is a list of URLs from which the Thing Descriptions can be retrieved, or the content of the Thing Descriptions themselves.

**Comment by farshidtz** [comment]
Some things have been discussed in https://github.com/w3c/wot-discovery/issues/18.

**Comment by mmccool** [comment]
@benfrancis we do have some specific issues where adding information to the TD is troublesome:
1. Signed TDs.  Sure, we can use proof chaining, but then the consumer has to trust the directory as well as the endpoint.  If we just want to add metadata to a TD (like the TTL of the TD *in this particular directory*) then it seems like we don't want to break signing for that.
2. Encrypted TDs.  It's possible a directory might want to store and transmit a TD upon request that it can't itself view.  Of course in this case the consumer would need to know the key used to look it up (during registration the registering entity would have to provide a "key", which may or may not be the same as the ID inside the TD, which the directory can't access...).   BTW by key I don't mean the decryption key, I just mean some value used to find the encrypted TD in the directory.
3. Geolocation.  I've been thinking a lot about how to deal with geolocation queries.  This information can be encoded in the TD (for static location, e.g. infrastructure devices) or can be provided by a property (for dynamic location, e.g. mobile devices).  It might also be inferred by the directory (e.g. within WiFi range).   So we could store the directories best guess as the location of the Thing in metadata and use that for queries.   Then all the geolocation is in one unified place, making setting up queries easier.

**Comment by mmccool** [comment]
Anyhow, another issue is that IDs are optional in TDs.  So yeah, the internal ID would be a useful key except we need to deal with TDs that omit them.

**Comment by benfrancis** [comment]
I agree that it would make sense to include relatively static geolocation metadata inside a Thing Description using JSON-LD (e.g. using the schema.org [GeoCoordinates](https://schema.org/GeoCoordinates) schema), but dynamic location data might make sense as a property (e.g. for a device which contains a GPS receiver).

[Encrypting](https://tools.ietf.org/html/rfc5246) and [signing](https://tools.ietf.org/id/draft-cavage-http-signatures-12.html) resources sound to me like two more things that the World Wide Web can already do that shouldn't need re-inventing for the Web of Things. Does someone have a real world use case for a directory which can't read the Thing Descriptions it serves, or is that just a hypothetical example? If this is a requirement then it may be necessary to define a completely separate MIME type for encrypted TDs and the TD would presumably no longer be indexable/queryable/searchable via a directory service, only retrievable by its URL.

I still maintain that the URI of the Thing Description makes the best identifier, and certainly a good fallback identifier if no ID is provided inside the Thing Description.

**Comment by farshidtz** [comment]
From the Discovery call:
- This is already covered in the specification.
